Modules/CarouselView/CarouselView.swift

```diff
@@ -0,0 +1,187 @@
+/// ```
+public final class CarouselView<ContentView: StatefulViewProtocol>: StatefulView<CarouselViewModel<ContentView.Model>>, UIScrollViewDelegate {
+    private lazy var scrollView: UIScrollView = .instantiate()
+    private lazy var contentViews: [ContentView] = []
+    private lazy var leftContentViewPlaceholder: ContentView = .instantiate()
+    private lazy var rightContentViewPlaceholder: ContentView = .instantiate()
+    private lazy var leftOffscreenContentViewPlaceholder: ContentView = .instantiate()
+    private lazy var rightOffscreenContentViewPlaceholder: ContentView = .instantiate()
+
+    private var contentViewIndex: Int = 0
+    private var isScrollViewDidScrollObservationEnabled: Bool = true
+
+    private var allContentViews: [ContentView] {
+        return [leftContentViewPlaceholder] + contentViews + [rightContentViewPlaceholder]
+    }
+
+    public override func viewDidLoad() {
+        super.viewDidLoad()
+
+        scrollView.delegate = self
+        scrollView.isPagingEnabled = true
+        scrollView.showsHorizontalScrollIndicator = false
+        scrollView.showsVerticalScrollIndicator = false
+        scrollView.translatesAutoresizingMaskIntoConstraints = false
+        addSubview(scrollView)
+        scrollView.topAnchor.constraint(equalTo: topAnchor).isActive = true
+        scrollView.leadingAnchor.constraint(equalTo: leadingAnchor).isActive = true
+        scrollView.trailingAnchor.constraint(equalTo: trailingAnchor).isActive = true
+        scrollView.bottomAnchor.constraint(equalTo: bottomAnchor).isActive = true
+        scrollView.addSubview(leftContentViewPlaceholder)
+        scrollView.addSubview(rightContentViewPlaceholder)
+        scrollView.addSubview(leftOffscreenContentViewPlaceholder)
+        scrollView.addSubview(rightOffscreenContentViewPlaceholder)
+    }
+
+    public override func didChangeModel() {
+        super.didChangeModel()
+
+        scrollView.alwaysBounceHorizontal = true
+        scrollView.alwaysBounceVertical = false
+
+        if model.pages.count < contentViews.count {
+            let removedIndices = model.pages.count ..< contentViews.count
+            contentViews[removedIndices].forEach { $0.removeFromSuperview() }
+            contentViews.removeSubrange(removedIndices)
+        } else if contentViews.count < model.pages.count {
+            let addedIndices = contentViews.count ..< model.pages.count
+            addedIndices.forEach { _ in
+                let view: ContentView = .instantiate()
+                contentViews.append(view)
+                scrollView.addSubview(view)
+            }
+        }
+
+        contentViews.enumerated().forEach { index, view in
+            view.model = model.pages[index]
+        }
+
+        if !model.pages.isEmpty {
+            leftContentViewPlaceholder.model = model.pages[modIndex(model.pages.count - 1, model.pages.count)]
+            rightContentViewPlaceholder.model = model.pages[0]
```

Comment:
Are you sure about index zero here? Shouldn't zero be the one cantered? Doesn't it have to be flexible?

Reply:
Yes that is correct here we do add the first page to the end of the list